library/aes.c

```diff
@@ -33,6 +33,31 @@
 #include "mbedtls/platform.h"
 #include "mbedtls/platform_util.h"
 #include "mbedtls/error.h"
+
+#if defined(__aarch64__)
+#if !defined(MBEDTLS_AESCE_C) && defined(MBEDTLS_AES_USE_HARDWARE_ONLY)
+#error "MBEDTLS_AES_USE_HARDWARE_ONLY defined, but not all prerequisites"
+#endif
+#endif
+
+#if defined(__amd64__) || defined(__x86_64__) || \
+    defined(_M_X64) || defined(_M_AMD64)
+#if !defined(MBEDTLS_AESNI_C) && defined(MBEDTLS_AES_USE_HARDWARE_ONLY)
+#error "MBEDTLS_AES_USE_HARDWARE_ONLY defined, but not all prerequisites"
+#endif
+#endif
+
+#if defined(__i386__) || defined(_M_IX86)
+#if defined(MBEDTLS_AES_USE_HARDWARE_ONLY) && !defined(MBEDTLS_AESNI_C)
+#error "MBEDTLS_AES_USE_HARDWARE_ONLY defined, but not all prerequisites"
+#endif
+
+#if defined(MBEDTLS_PADLOCK_C) && \
+    (!defined(MBEDTLS_HAVE_ASM) || defined(MBEDTLS_AES_USE_HARDWARE_ONLY))
+#error "MBEDTLS_PADLOCK_C defined, but not all prerequisites"
+#endif
```

Comment:
We don't support `MBEDTLS_AES_USE_HARDWARE_ONLY` with `MBEDTLS_PADLOCK_C`, which means users MUST explicitly unset `MBEDTLS_PADLOCK_C` based on the default configuration if they want to use `MBEDTLS_AES_USE_HARDWARE_ONLY`. In that case, the error message would be confusing. 
How about:
```C
#if defined(MBEDTLS_PADLOCK_C)
#if !defined(MBEDTLS_HAVE_ASM)
#error "MBEDTLS_PADLOCK_C defined, but not all prerequisites"
#endif
#if defined(MBEDTLS_AES_USE_HARDWARE_ONLY)
#error "MBEDTLS_AES_USE_HARDWARE_ONLY defined, but it does not support MBEDTLS_PADLOCK_C"
#endif
#endif
```

Reply:
The existing message in `check_config.h` for when we have two incompatible defines is

```
#error "MBEDTLS_HAVE_INT32 and MBEDTLS_HAVE_INT64 cannot be defined simultaneously" 
```

but here I would suggest

```
#error "MBEDTLS_AES_USE_HARDWARE_ONLY cannot be defined when MBEDTLS_PADLOCK_C is set"
```

(HAVE_INT32/64 are symmetric, whereas HARDWARE_ONLY and PADLOCK are not, and phrasing this way implies that unsetting PADLOCK is a reasonable thing to do)